lucille-plugins/lucille-pinecone/src/test/java/com/kmwllc/lucille/pinecone/indexer/PineconeIndexerTest.java

```diff
@@ -0,0 +1,348 @@
+      indexerGood.validateConnection();
+      indexerGood.closeConnection();
+      Mockito.verify(goodConnection, Mockito.times(1)).close();
+    }
+  }
+
+
+  @Test
+  public void testUpsertAndUpdateNoNamespacesProvided() throws Exception {
+    try (MockedConstruction<PineconeClient> client = Mockito.mockConstruction(PineconeClient.class, (mock, context) -> {
+      Mockito.when(mock.connect("good")).thenReturn(goodConnection);
+    })) {
+
+      TestMessenger messenger = new TestMessenger();
+      TestMessenger messenger2 = new TestMessenger();
+      Config configUpsert = ConfigFactory.load("PineconeIndexerTest/no-namespaces.conf");
+      Config configUpdate = ConfigFactory.load("PineconeIndexerTest/no-namespaces-update.conf");
+
+      PineconeIndexer indexerUpdate = new PineconeIndexer(configUpdate, messenger, "testing");
+      PineconeIndexer indexerUpsert = new PineconeIndexer(configUpsert, messenger2, "testing");
+
+      indexerUpsert.validateConnection();
+      indexerUpdate.validateConnection();
+
+      messenger.sendForIndexing(doc0);
+      messenger.sendForIndexing(doc1);
+      messenger2.sendForIndexing(doc0);
+      messenger2.sendForIndexing(doc1);
+      indexerUpdate.run(2);
+      indexerUpsert.run(2);
+
+      // assert that no upserts or updates have happened
+      Mockito.verify(stub, Mockito.times(0)).update(Mockito.any());
+      Mockito.verify(stub, Mockito.times(0)).upsert(Mockito.any());
+    }
+  }
+
+
+  @Test
+  public void testUpsertMultipleNamespaces() throws Exception {
+    try (MockedConstruction<PineconeClient> client = Mockito.mockConstruction(PineconeClient.class, (mock, context) -> {
+      Mockito.when(mock.connect("good")).thenReturn(goodConnection);
+    })) {
+
+      TestMessenger messenger = new TestMessenger();
+      Config configGood = ConfigFactory.load("PineconeIndexerTest/two-namespaces.conf");
+      PineconeIndexer indexerGood = new PineconeIndexer(configGood, messenger, "testing");
+      indexerGood.validateConnection();
+
+      messenger.sendForIndexing(doc0);
+      messenger.sendForIndexing(doc1);
+      indexerGood.run(2);
+
+      // make sure no updates were made
+      ArgumentCaptor<UpdateRequest> updateRequest = ArgumentCaptor.forClass(UpdateRequest.class);
+      Mockito.verify(stub, Mockito.times(0)).update(Mockito.any());
+      // make sure two upserts were made
+      ArgumentCaptor<UpsertRequest> upsertRequest = ArgumentCaptor.forClass(UpsertRequest.class);
+      Mockito.verify(stub, Mockito.times(2)).upsert(upsertRequest.capture());
+
+      UpsertRequest namespace1Upsert = upsertRequest.getAllValues().get(0);
+      UpsertRequest namespace2Upsert = upsertRequest.getAllValues().get(1);
+
+      assertEquals("namespace-1", namespace1Upsert.getNamespace());
+      assertEquals("namespace-2", namespace2Upsert.getNamespace());
+
+      assertEquals(2, namespace1Upsert.getVectorsList().size());
+      assertEquals(2, namespace2Upsert.getVectorsList().size());
+
```

Comment:
Can remove this extra space. There are a few other places to do this throughout the file.

Reply:
done